docs/basics/upgradeability.md

```diff
@@ -122,6 +131,172 @@ Using this pattern, you can introduce other message to your proxy contract.
 Any messages that are not matched in the proxy contract 
 will be forwarded to the specified contract address.
 
+## Delegating execution to foreign Contract Code with `delegate_call`
+
+Similar to proxy-forwarding we can delegate execution to another code hash uploaded on-chain.
+
+### Properties 
+
+- Delegates any call that does not match a selector of itself to another contract.
+- Code is required to uploaded on-chain, but is not required to be instantiated.
+- State is stored in the storage of the original contract which submits the call.
+- Storage layout must be identical between both contract codes.
+
+```
+User ---- tx ---> Contract ----------> Code_v0
+                     ^                     |
+                     |                     |
+                     ⌊_____________________⌋
+                  Storage update is propagated
+```
+
+### Example
+
+Suppose we have the defined of the caller contract as following:
+
+```rust
+#[ink(storage)]
+pub struct Delegator {
+    addresses: Mapping<AccountId, i32, ManualKey<0x23>>,
+    counter: i32,
+}
+```
+
+Then let's define two messages that separately calls to update `addresses` and `counter` separately:
+
+```rust
+/// Increment the current value using delegate call.
+#[ink(message)]
+pub fn inc_delegate(&mut self, hash: Hash) {
+    let selector = ink::selector_bytes!("inc");
+    let _ = build_call::<DefaultEnvironment>()
+        .delegate(hash)
+        // We specify `set_tail_call(true)` to use the delegatee last memory frame
+        // as the end of the execution cycle.
+        // So any mutations to `Packed` types, made by delegatee,
+        // will be flushed to storage.
+        //
+        // If we don't specify this flag.
+        // The storage state before the delegate call will be flushed to storage instead.
+        // See https://substrate.stackexchange.com/questions/3336/i-found-set-allow-reentry-may-have-some-problems/3352#3352
+        .call_flags(CallFlags::default().set_tail_call(true))
+        .exec_input(ExecutionInput::new(Selector::new(selector)))
+        .returns::<()>()
+        .try_invoke();
+}
+
+/// Adds entry to `addresses` using delegate call.
+/// Note that we don't need `set_tail_call(true)` flag
+/// because `Mapping` updates the storage instantly on-demand.
+#[ink(message)]
+pub fn add_entry_delegate(&mut self, hash: Hash) {
+    let selector = ink::selector_bytes!("append_address_value");
+    let _ = build_call::<DefaultEnvironment>()
```

Comment:
Perhaps we could demo or mention using the `ContractRef` methods to initialize the call builder?

Reply:
I don't recal `ContractRef` to be used for initialization of call builder, rather create builder.

Reply:
You can use it for call builder, but I realise now we can't use it for `delegate` calls as it only generates [`call` or `call_mut`](https://github.com/paritytech/ink/blob/bf021b0efc71d8e27b13367e7ea0ab3cea406542/crates/ink/codegen/src/generator/as_dependency/contract_ref.rs#L382).